File: message_sync/src/sync_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "std_msgs/msg/int32.hpp"
#include "message_filters/subscriber.h"
#include "message_filters/synchronizer.h"
#include "message_filters/sync_policies/approximate_time.h"

typedef message_filters::sync_policies::ApproximateTime<std_msgs::msg::String, std_msgs::msg::Int32> SyncPolicy;
typedef message_filters::Synchronizer<SyncPolicy> Synchronizer;

class SyncSubscriber : public rclcpp::Node
{
public:
    SyncSubscriber() : Node("sync_subscriber")
    {
        sub1_ = std::make_shared<message_filters::Subscriber<std_msgs::msg::String>>(
            this, "camera_data"
        );

        sub2_ = std::make_shared<message_filters::Subscriber<std_msgs::msg::Int32>>(
            this, "lidar_data"
        );

        sync_ = std::make_shared<Synchronizer>(SyncPolicy(10), *sub1_, *sub2_);

        sync_->registerCallback(
            std::bind(&SyncSubscriber::syncCallback, this,
                std::placeholders::_1,
                std::placeholders::_2)
        );

        RCLCPP_INFO(this->get_logger(), "Sync Subscriber started !");
    }

private:
    void syncCallback(
        const std_msgs::msg::String::ConstSharedPtr msg1,
        const std_msgs::msg::Int32::ConstSharedPtr msg2)
    {
        RCLCPP_INFO(this->get_logger(),
            "SYNCED → camera: '%s' | lidar: %d",
            msg1->data.c_str(), msg2->data);
    }

    std::shared_ptr<message_filters::Subscriber<std_msgs::msg::String>> sub1_;
    std::shared_ptr<message_filters::Subscriber<std_msgs::msg::Int32>>  sub2_;
    std::shared_ptr<Synchronizer> sync_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SyncSubscriber>());
    rclcpp::shutdown();
    return 0;
}

File: message_sync_py/message_sync_py/sync_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String, Int32

# ============================================================
# message_filters en Python
# ============================================================
import message_filters

class SyncSubscriber(Node):
    def __init__(self):
        super().__init__("sync_subscriber")

        # --------------------------------------------------------
        # CRÉER LES SUBSCRIBERS avec message_filters
        # --------------------------------------------------------
        self.sub1_ = message_filters.Subscriber(self, String, "camera_data")
        self.sub2_ = message_filters.Subscriber(self, Int32,  "lidar_data")

        # --------------------------------------------------------
        # CRÉER LE SYNCHRONIZER
        # --------------------------------------------------------
        self.sync_ = message_filters.ApproximateTimeSynchronizer(
            [self.sub1_, self.sub2_],   # liste des subscribers
            queue_size=10,              # messages en attente
            slop=0.1                    # tolérance en secondes
            # slop = écart maximum accepté entre les timestamps
        )

        # Enregistrer le callback
        self.sync_.registerCallback(self.sync_callback)

        self.get_logger().info("Sync Subscriber started !")

    # --------------------------------------------------------
    # CALLBACK DE SYNCHRONISATION
    # --------------------------------------------------------
    def sync_callback(self, msg1, msg2):
        self.get_logger().info(
            f"SYNCED → camera: '{msg1.data}' | lidar: {msg2.data}"
        )

def main(args=None):
    rclpy.init(args=args)
    node = SyncSubscriber()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()